pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,520 @@
+	}
+
+	// if the binding resolution is not yet noted - create a new entry
+	objectIdentifierToResolution, exists := c.bindingNameToResolutions[bindingName]
+	if !exists {
+		objectIdentifierToResolution = make(map[util.ObjectIdentifier]*combinedStatusResolution)
+		c.bindingNameToResolutions[bindingName] = objectIdentifierToResolution
+	}
+
+	// (2) remove excessive combinedstatus resolutions of objects that are no longer
+	// associated with the binding resolution
+	for objectIdentifier, _ := range objectIdentifierToResolution {
+		if _, exists := bindingResolution.ObjectIdentifierToData[objectIdentifier]; !exists {
+			combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, objectIdentifier))
+			delete(objectIdentifierToResolution, objectIdentifier)
+		}
+	}
+
+	// (~2+3) create/update combinedstatus resolutions for every object that requires status collection,
+	// and delete resolutions that are no longer required
+	for objectIdentifier, objectData := range bindingResolution.ObjectIdentifierToData {
+		csResolution, exists := objectIdentifierToResolution[objectIdentifier]
+		if len(objectData.StatusCollectors) == 0 {
+			if exists { // associated resolution is no longer required
+				combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, objectIdentifier))
+				delete(objectIdentifierToResolution, objectIdentifier)
+			}
+
+			continue
+		}
+
+		// create resolution entry if missing
+		if !exists {
+			csResolution = &combinedStatusResolution{
+				statusCollectorNameToData: make(map[string]*statusCollectorData),
+			}
+			objectIdentifierToResolution[objectIdentifier] = csResolution
+		}
+
+		// fetch missing statuscollector specs
+		c.fetchMissingStatusCollectorSpecsLocked(statusCollectorLister, objectData.StatusCollectors)
+
+		// update statuscollectors
+		removedCollectors, addedCollectors := csResolution.setStatusCollectors(
+			abstract.SliceToPrimitiveMap(objectData.StatusCollectors,
+				func(statusCollectorName string) string { return statusCollectorName }, // keys mapper
+				func(statusCollectorName string) *v1alpha1.StatusCollectorSpec { // val mapper
+					return c.statusCollectorNameToSpec[statusCollectorName] // if missing after fetch, ignored
+				}))
+
+		// update destinations
+		removedDestinations, newDestinationsSet := csResolution.setCollectionDestinations(collectionDestinations)
+
+		// should queue the combinedstatus object for syncing if lost collectors / destinations
+		if removedCollectors || removedDestinations {
+			combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, objectIdentifier))
+		}
+
+		// should evaluate workstatuses if added/updated collectors or added destinations
+		// it's possible to evaluate workloads coming from new destinations only if no collectors were removed (TODO)
+		if addedCollectors || len(newDestinationsSet) > 0 {
+			workloadIdentifiersToEvaluate.Insert(objectIdentifier)
```

Comment:
The current approach here will do more work than necessary. To explain, let me define the following symbols. All of the remarks here are scoped to one workload object (one run through the body of the loop over `bindingResolution.ObjectIdentifierToData`).

- `collectorsToRemove`, the ones that were removed in `csResolution.setStatusCollectors`
- `collectorsToKeep`, the ones that were already in the domain of `csResolution.statusCollectorNameToData` at the start of this func AND are listed in `objectData.StatusCollectors`.
- `collectorsToAdd`, the ones listed in `objectData.StatusCollectors` BUT NOT already in the domain of  `csResolution.statusCollectorNameToData` at the start of this func.

- `destinationsToRemove`, the ones that had data in `csResolution` at the start of this func but are NOT listed in `bindingResolution.Destinations`.
- `destinationsToKeep`, the ones that had data in `csResolution` at the start of this func AND ARE listed in `bindingResolution.Destinations.
- `destinationsToAdd`, the ones that are in `bindingResolution.Destinations` BUT did NOT have data in `csResolution` at the start of this func.

At the start of this func, csResolution had materialized view rows for the cross product of (collectorsToRemove U collectorsToKeep) X (destinationsToRemove U destinationsToKeep) (here I use "U" for set union). This func needs to change csResolution so that it has rows for the cross product of (collectorsToKeep U collectorsToAdd) X (destinationsToKeep U destinationsToAdd). At this point in the code, three blocks has already been removed:

- collectorsToRemove X destinationsToRemove,
- collectorsToRemove X destinationsToKeep, and
- collectorsToKeep X desinationsToRemove

The rows in the block collectorsToKeep X destinationsToKeep have been left untouched and should remain. Because this func is responding to a notifcation about a state change in a binding resolution rather than a workload object or a StatusCollector, the rows in collectorsToKeep X destinationsToKeep do not need to be re-evaluated.

What remains to be done is add rows for three blocks:

- collectorsToAdd X destinationsToAdd,
- collectorsToAdd X destinationsToKeep, and
- collectorsToKeep X destinationsToAdd.

These will be new rows, and comparison with previous contents does not need to be done.

Reply:
Thank you for putting it cleanly. I am aware of the excessive work done here, will indeed be tightened in a follow-up PR.